sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -441,6 +537,17 @@ class Dataset[T] private(
   /** Collects the elements to an Array. */
   def collect(): Array[T] = rdd.collect()
 
+  /**
+   * (Java-specific)
+   * Collects the elements to a Java list.
+   *
+   * Due to the incompatibility problem between Scala and Java, the return type of [[collect()]] at
```

Comment:
We can construct right type of array while calling `RDD.collect`, the problem is the interface. At java side the return type of `RDD.collect()` is `java.lang.Object` and we need to do a type cast, which is not friendly to users.